vignettes/deprecations.Rmd

```diff
@@ -0,0 +1,131 @@
+---
+title: "Handling deprecated Stan features with the canonicalizer in CmdStanR"
+author: "Rok Češnovar, Jonah Gabry"
+output: 
+  rmarkdown::html_vignette:
+    toc: true
+    toc_depth: 4
+resource_files:
+   - deprecations-files/logistic.stan
+params:
+  EVAL: !r identical(Sys.getenv("NOT_CRAN"), "true")
+vignette: >
+  %\VignetteIndexEntry{Handling deprecated Stan features with the canonicalizer in CmdStanR}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+```{r child="children/settings-knitr.Rmd"}
+```
+
+## Introduction
+
+This vignette demonstrates how to handle cases where one or more feature
+in your Stan model has been deprecated. For most deprecated features, the
+Stan-to-C++ compiler can be used to automatically convert to a non-deprecated
+feature that replaces the deprecated one. This vignette show cases how that
+automatic conversion can be done using CmdStanR.
+
+The automatic conversion of deprecated features to non-deprecated features
+is done by with a mode called `canonicalizer`, which is part of the Stan-to-C++
+compiler. We recommend using CmdStan 2.29.2 or newer when using the canonicalizer
+and this vignette. The minimum CmdStanR version to run the vignette is 0.5.0.
+
+```{r library, message=FALSE}
+library(cmdstanr)
+check_cmdstan_toolchain(fix = TRUE, quiet = TRUE)
+```
+
+## Deprecation warnings
+
+Consider the following simple logistic regression with parameters `alpha` and `beta`,
+covariates `X`, and outcome `y`:
+
+```stan
+data {
+  int<lower=1> k;
+  int<lower=0> n;
+  matrix[n, k] X;
+  int y[n];
+}
+parameters {
+  vector[k] beta;
+  real alpha;
+}
+model {
+  # priors
+  target += std_normal_log(beta);
+  alpha ~ std_normal();
+  
+  y ~ bernoulli_logit(X * beta + alpha);
+}
+```
+
+When compiling the model we receive a few warnings:
+
+```{r compile, message=FALSE}
+mod <- cmdstan_model(stan_file = system.file("deprecated-logistic.stan", package = "cmdstanr"))
```

Comment:
I had to put the file in the `inst` folder for it to pass tests here. If I put it in a subfolder of vignettes, it was unable to find it (building articles locally does work).

Reply:
I would prefer this model being in the vignettes subfolder though, but have no idea if that is doable.

Reply:
I can look into that

Reply:
Actually in this case I think we should just put the Stan program directly in the vignette and use `write_stan_file()`. That way we can also demonstrate the `overwrite_file` feature instead of having to set eval=FALSE for that. I can update the vignette

Reply:
If you don't like using `write_stan_file` here I can switch it back

Reply:
Ah yeah, we can use write_stan_file. I try to avoid it, to not write the model in a string, but for this purpose, it is fine.